src/Backend/ONE/ONEToolchain.ts

```diff
@@ -14,33 +14,106 @@
  * limitations under the License.
  */
 
+import {execSync} from 'child_process';
 import {Backend} from '../Backend';
 import {Command} from '../Command';
 import {Compiler} from '../Compiler';
 import {Executor} from '../Executor';
-import {Toolchains} from '../Toolchain';
+import {ToolchainInfo, Toolchains} from '../Toolchain';
+import {DebianArch, DebianRepo, DebianToolchain} from '../ToolchainImpl/DebianToolchain';
+import {Version} from '../Version';
 
 class ToolchainCompiler implements Compiler {
   private readonly toolchainTypes: string[];
+  private readonly toolchainName: string;
+  private readonly debianRepo: DebianRepo;
+  private readonly debianArch: DebianArch;
 
   constructor() {
-    this.toolchainTypes = [];
+    this.toolchainName = 'onecc-docker-test';
+    // we currently support only the latest version of one-compiler.
+    this.toolchainTypes = ['latest'];
+
+    // we temporary add personal PPA for test.
+    this.debianRepo =
+        new DebianRepo('http://ppa.launchpad.net/psalmist-kim/ppa/ubuntu', 'bionic', 'main');
+    this.debianArch = DebianArch.amd64;
   }
 
   getToolchainTypes(): string[] {
     return this.toolchainTypes;
   }
 
-  getToolchains(_toolchainType: string, _start: number, _count: number): Toolchains {
-    throw Error('Invalid getToolchains call');
+  getToolchains(_toolchainType: string, _start?: number, _count?: number): Toolchains {
+    if (parseInt(execSync(`apt-cache show ${this.toolchainName} > /dev/null 2>&1; echo $?`)
+                     .toString()
+                     .trim()) !== 0) {
+      throw Error('the prerequisites are not met');
+    }
+
+    const availableToolchains = new Toolchains();
+
+    if (_toolchainType === 'latest') {
+      const versionList =
+          execSync(`apt-cache show ${this.toolchainName} | grep Version | awk '{print $2}'`)
+              .toString()
+              .trim()
+              .split('\n');
+
+      for (const version of versionList) {
+        const [major, minor, patch, option] = version.split(/[.~]+/);
+
+        const toolchainInfo = new ToolchainInfo(
+            this.toolchainName, 'Description: test',
+            new Version(parseInt(major), parseInt(minor), parseInt(patch), option));
+
+        const toolchain = new DebianToolchain(toolchainInfo, this.debianRepo, this.debianArch);
+        availableToolchains.push(toolchain);
+      }
+    }
+
+    return availableToolchains;
   }
 
   getInstalledToolchains(_toolchainType: string): Toolchains {
-    throw Error('Invalid getInstalledToolchains call');
+    const availableToolchains = new Toolchains();
```

Comment:
```    
const installedToolchains = this.getToolchains(_toolchainType).filter((toolchain) => {
      return parseInt(execSync(`dpkg-query --show ${toolchain.info.name} > /dev/null 2>&1; echo $?`)
                          .toString()
                          .trim()) === 0;
    });
    return installedToolchains;
```

At first we wrote code like this. But it fails the test. We assume the cause of the fail is calling `getToolchains` in `getInstalledToolchains`. So we wrote some duplicate codes for passing the test. 

Could you give us some advice to improve this code?